src/dafny/BinaryOperations.dfy

```diff
@@ -0,0 +1,202 @@
+// RUN: %verify "%s"
+
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT 
+ *******************************************************************************/
+
+module {:options "-functionSyntax:4"} BinaryOperations {
+
+  ghost predicate IsAssociative<T(!new)>(bop: (T, T) -> T) {
+    forall x, y, z :: bop(bop(x, y), z) == bop(x, bop(y, z))
+  }
+
+  ghost predicate IsLeftUnital<T(!new)>(bop: (T, T) -> T, unit: T) {
+    forall x :: bop(unit, x) == x
+  }
+
+  ghost predicate IsRightUnital<T(!new)>(bop: (T, T) -> T, unit: T) {
+    forall x :: bop(x, unit) == x
+  }
+
+  ghost predicate IsUnital<T(!new)>(bop: (T, T) -> T, unit: T) {
+    && IsLeftUnital(bop, unit)
+    && IsRightUnital(bop, unit)
+  }
+
+  ghost predicate IsLeftInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    forall x | inverse.requires(x) :: bop(inverse(x), x) == unit
+  }
+
+  ghost predicate IsRightInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    forall x | inverse.requires(x) :: bop(x, inverse(x)) == unit
+  }
+
+  ghost predicate IsInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    && IsLeftInverse(bop, inverse, unit)
+    && IsRightInverse(bop, inverse, unit)
+  }
+
+  ghost predicate IsAbelian<T(!new)>(bop: (T, T) -> T) {
+    forall x, y :: bop(x, y) == bop(y, x)
+  }
+
+  ghost predicate IsRightDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    forall x, y, z :: bop2(bop1(x, y), z) == bop1(bop2(x, z), bop2(y, z))
+  }
+
+  ghost predicate IsLeftDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    forall x, y, z :: bop2(x, bop1(y, z)) == bop1(bop2(x, y), bop2(x, z))
+  }
+
+  ghost predicate IsDistributive<T(!new)>(bop1: (T, T) -> T, bop2: (T, T) -> T) {
+    && IsLeftDistributive(bop1, bop2)
+    && IsRightDistributive(bop1, bop2)
+  }
+
+  ghost predicate IsMonoid<T(!new)>(bop: (T, T) -> T, unit: T) {
+    && IsAssociative(bop)
+    && IsUnital(bop, unit)
+  }
+
+  datatype MonoidStructure<!T(!new)> = MonoidStructure(bop: (T, T) -> T, unit: T)
+  type Monoid<!T(!new)> = g: MonoidStructure | IsMonoid(g.bop, g.unit) witness *
```

Comment:
Could we use, say ((x,y)->x*y,1) as a witness?
And something similar for the other witness *

Reply:
I was thinking about providing `MonoidStructure((x,y) => c, c)` as witness, where `c` is any element in `T`. If we assume a non-empty type `T(00)`, we indeed can deduce an element `c`, in a ghost context. However, unfortunately, to use it to construct a monoid witness, it needs to be non-ghost, i.e. uniquely determined. For now, I don't think we can make this assumption.